File: js/src/result.ts
/** Streaming result cursor. */

import type { ClientReadableStream } from "@grpc/grpc-js";
import type { ExecuteResponse } from "./generated/gql_service";
import { valueFromProto } from "./convert";
import { isSuccess } from "./status";
import type { GqlValue } from "./types";

/** Summary of a completed query execution. */
export class ResultSummary {
  /** @internal */
  constructor(private readonly proto: NonNullable<ExecuteResponse["summary"]>) {}

  /** The GQLSTATUS code. */
  get statusCode(): string {
    return this.proto.status?.code ?? "";
  }

  /** The status message. */
  get message(): string {
    return this.proto.status?.message ?? "";
  }

  /** Number of rows affected. */
  get rowsAffected(): bigint {
    return this.proto.rowsAffected;
  }

  /** Operation counters. */
  get counters(): ReadonlyMap<string, bigint> {
    return new Map(Object.entries(this.proto.counters));
  }

  /** Check if the execution was successful. */
  isSuccess(): boolean {
    return isSuccess(this.statusCode);
  }
}

/** Cursor over streaming result frames from an Execute RPC. */
export class ResultCursor implements AsyncIterable<GqlValue[]> {
  private readonly stream: ClientReadableStream<ExecuteResponse>;
  private header: ExecuteResponse["header"] | undefined;
  private _summary: ExecuteResponse["summary"] | undefined;
  private bufferedRows: GqlValue[][] = [];
  private rowIndex = 0;
  private done = false;
  private pendingResolve: (() => void) | null = null;

  /** @internal */
  constructor(stream: ClientReadableStream<ExecuteResponse>) {
    this.stream = stream;

    stream.on("data", (response: ExecuteResponse) => {
      if (response.header) {
        this.header = response.header;
      } else if (response.rowBatch) {
        for (const row of response.rowBatch.rows) {
          this.bufferedRows.push(row.values.map(valueFromProto));
        }
      } else if (response.summary) {
        this._summary = response.summary;
        this.done = true;
      }

      if (this.pendingResolve) {
        const resolve = this.pendingResolve;
        this.pendingResolve = null;
        resolve();
      }
    });

    stream.on("end", () => {
      this.done = true;
      if (this.pendingResolve) {
        const resolve = this.pendingResolve;
        this.pendingResolve = null;
        resolve();
      }
    });

    stream.on("error", () => {
      this.done = true;
      if (this.pendingResolve) {
        const resolve = this.pendingResolve;
        this.pendingResolve = null;
        resolve();
      }
    });
  }

  private waitForData(): Promise<void> {
    if (this.done || this.rowIndex < this.bufferedRows.length) {
      return Promise.resolve();
    }
    return new Promise<void>((resolve) => {
      this.pendingResolve = () => resolve();
    });
  }

  private waitForDone(): Promise<void> {
    if (this.done) return Promise.resolve();
    return new Promise<void>((resolve) => {
      this.pendingResolve = () => {
        if (this.done) resolve();
        else this.waitForDone().then(resolve);
      };
    });
  }

  /** Get the column names from the result header. */
  async columnNames(): Promise<string[]> {
    while (!this.header && !this.done) {
      await this.waitForData();
    }
    return this.header?.columns.map((c) => c.name) ?? [];
  }

  /** Get the next row. Returns null when done. */
  async nextRow(): Promise<GqlValue[] | null> {
    if (this.rowIndex < this.bufferedRows.length) {
      return this.bufferedRows[this.rowIndex++];
    }

    while (!this.done) {
      await this.waitForData();
      if (this.rowIndex < this.bufferedRows.length) {
        return this.bufferedRows[this.rowIndex++];
      }
    }

    return null;
  }

  /** Collect all remaining rows. */
  async collectRows(): Promise<GqlValue[][]> {
    const rows: GqlValue[][] = [];
    let row = await this.nextRow();
    while (row !== null) {
      rows.push(row);
      row = await this.nextRow();
    }
    return rows;
  }

  /** Get the result summary. Consumes remaining frames if needed. */
  async summary(): Promise<ResultSummary | null> {
    await this.waitForDone();
    return this._summary ? new ResultSummary(this._summary) : null;
  }

  /** Check if the execution was successful. */
  async isSuccess(): Promise<boolean> {
    const s = await this.summary();
    return s?.isSuccess() ?? false;
  }

  /** Get the number of rows affected. */
  async rowsAffected(): Promise<bigint> {
    const s = await this.summary();
    return s?.rowsAffected ?? 0n;
  }

  async *[Symbol.asyncIterator](): AsyncIterator<GqlValue[]> {
    let row = await this.nextRow();
    while (row !== null) {
      yield row;
      row = await this.nextRow();
    }
  }
}

File: python/src/gwp_py/result.py
"""Streaming result cursor."""

from __future__ import annotations

from typing import Any

import grpc

from gwp_py._convert import value_from_proto
from gwp_py.status import is_success


class ResultSummary:
    """Summary of a completed query execution."""

    def __init__(self, proto_summary: Any):
        self._proto = proto_summary

    @property
    def status_code(self) -> str:
        """The GQLSTATUS code."""
        return self._proto.status.code if self._proto.status else ""

    @property
    def message(self) -> str:
        """The status message."""
        return self._proto.status.message if self._proto.status else ""

    @property
    def rows_affected(self) -> int:
        """Number of rows affected."""
        return self._proto.rows_affected

    @property
    def counters(self) -> dict[str, int]:
        """Operation counters."""
        return dict(self._proto.counters)

    def is_success(self) -> bool:
        """Check if the execution was successful."""
        return is_success(self.status_code)


class ResultCursor:
    """Cursor over streaming result frames from an Execute RPC."""

    def __init__(self, stream: Any):
        self._stream = stream
        self._header: Any | None = None
        self._summary: Any | None = None
        self._buffered_rows: list[list[Any]] = []
        self._row_index: int = 0
        self._done: bool = False

    async def _consume_until_rows_or_done(self) -> None:
        """Read frames until we have buffered rows or reach the end."""
        while not self._done and self._row_index >= len(self._buffered_rows):
            try:
                response = await self._stream.read()
                if response is grpc.aio.EOF:
                    self._done = True
                    return
            except Exception:
                self._done = True
                return

            frame = response.WhichOneof("frame")
            if frame == "header":
                self._header = response.header
            elif frame == "row_batch":
                for row in response.row_batch.rows:
                    self._buffered_rows.append(
                        [value_from_proto(v) for v in row.values]
                    )
            elif frame == "summary":
                self._summary = response.summary
                self._done = True

    async def column_names(self) -> list[str]:
        """Get the column names from the result header."""
        if self._header is None:
            await self._consume_until_rows_or_done()
        if self._header is None:
            return []
        return [col.name for col in self._header.columns]

    async def next_row(self) -> list[Any] | None:
        """Get the next row. Returns None when done."""
        if self._row_index < len(self._buffered_rows):
            row = self._buffered_rows[self._row_index]
            self._row_index += 1
            return row

        await self._consume_until_rows_or_done()

        if self._row_index < len(self._buffered_rows):
            row = self._buffered_rows[self._row_index]
            self._row_index += 1
            return row

        return None

    async def collect_rows(self) -> list[list[Any]]:
        """Collect all remaining rows."""
        rows: list[list[Any]] = []
        while True:
            row = await self.next_row()
            if row is None:
                break
            rows.append(row)
        return rows

    async def summary(self) -> ResultSummary | None:
        """Get the result summary. Consumes remaining frames if needed."""
        while not self._done:
            # Skip past any buffered rows so _consume can read more frames
            self._row_index = len(self._buffered_rows)
            await self._consume_until_rows_or_done()
        if self._summary is not None:
            return ResultSummary(self._summary)
        return None

    async def is_success(self) -> bool:
        """Check if the execution was successful."""
        s = await self.summary()
        return s.is_success() if s else False

    async def rows_affected(self) -> int:
        """Get the number of rows affected."""
        s = await self.summary()
        return s.rows_affected if s else 0

    def __aiter__(self) -> ResultCursor:
        return self

    async def __anext__(self) -> list[Any]:
        row = await self.next_row()
        if row is None:
            raise StopAsyncIteration
        return row
